Nodejs/Product/TestAdapter/TestExecutor.cs

```diff
@@ -75,148 +133,124 @@ class TestExecutor : ITestExecutor {
+        public void RunTests(IEnumerable<TestCase> tests, IRunContext runContext, IFrameworkHandle frameworkHandle)
+        {
             ValidateArg.NotNull(tests, "tests");
             ValidateArg.NotNull(runContext, "runContext");
             ValidateArg.NotNull(frameworkHandle, "frameworkHandle");
-
             _cancelRequested.Reset();
 
-            RunTestCases(tests, runContext, frameworkHandle);
+            // .ts file path -> project settings
+            var fileToTests = new Dictionary<string, List<TestCase>>();
+            var sourceToSettings = new Dictionary<string, NodejsProjectSettings>();
+            NodejsProjectSettings settings = null;
+
+            // put tests into dictionary where key is their source file
+            foreach (var test in tests)
+            {
+                if (!fileToTests.ContainsKey(test.CodeFilePath))
+                {
+                    fileToTests[test.CodeFilePath] = new List<TestCase>();
+                }
+                fileToTests[test.CodeFilePath].Add(test);
+            }
+
+            // where key is the file and value is a list of tests
+            foreach (KeyValuePair<string, List<TestCase>> entry in fileToTests)
+            {
+                TestCase firstTest = entry.Value.ElementAt(0);
+                if (!sourceToSettings.TryGetValue(firstTest.Source, out settings))
+                {
+                    sourceToSettings[firstTest.Source] = settings = LoadProjectSettings(firstTest.Source);
+                }
+
+                _currentTests = entry.Value;
+                _frameworkHandle = frameworkHandle;
+
+                // Run all test cases in a given file
+                RunTestCases(entry.Value, runContext, frameworkHandle, settings);
+            }
         }
 
-        private void RunTestCases(IEnumerable<TestCase> tests, IRunContext runContext, IFrameworkHandle frameworkHandle) {
+        private void RunTestCases(IEnumerable<TestCase> tests, IRunContext runContext, IFrameworkHandle frameworkHandle, NodejsProjectSettings settings) {
             // May be null, but this is handled by RunTestCase if it matters.
             // No VS instance just means no debugging, but everything else is
             // okay.
+            if (tests.Count() == 0)
+            {
+                return;
+            }
             using (var app = VisualStudioApp.FromEnvironmentVariable(NodejsConstants.NodeToolsProcessIdEnvironmentVariable)) {
+                int port = 0;
+                List<string> nodeArgs = new List<string>();
                 // .njsproj file path -> project settings
                 var sourceToSettings = new Dictionary<string, NodejsProjectSettings>();
+                List<TestCaseObject> testObjects = new List<TestCaseObject>();
+
+                if (!File.Exists(settings.NodeExePath))
+                {
+                    frameworkHandle.SendMessage(TestMessageLevel.Error, "Interpreter path does not exist: " + settings.NodeExePath);
```

Comment:
Agreed... but this assembly isn't localized at all currently, so that's a bigger effort.